### PR TITLE
fix: add extra logging for api keys request.

### PR DESCRIPTION
**Pull request by navdeepg2021**
Since we are getting 501 errors for the request to fetch the secret api keys/user identity, added some changes to log the request and response. 

This change should not add the logs unless the log level is configured to be debug or trace.

### CODE REVIEW

File: src/apis/source.rs [hunk 1]
```diff
@@ -460,17 +461,20 @@ impl SourceAPI {
...
                 source_api_url, access_key_id
             ))
             .headers(headers)
-            .send()
-            .await
-        {
+            .build()
+            .unwrap();
+        // Log the request details
+        debug!("request for api-keys: {:?}", request);
```
Comment (kbgg) [inline src/apis/source.rs:473]: Would this not output to the logs the sensitive `Authorization` header? That really would not be great from a security standpoint.

**Review by kbgg — COMMENTED** [review]
Hi Navdeep,

I have a couple of comments / concerns related to this PR:

1. From a security standpoint, logging the entire request including Auth headers to the logs is not great.
2. Have you been able to reproduce this issue in your local environment first? That should be your first step in debugging this issue. This PR is labeled as a "fix" and I don't see how adding more logs fixes this issue.
3. Your debugging seems to be occurring in a different project than the errors are happening in. From the screenshot that Amit shared with me (attached below) the errors are happening in the Vercel which is the [Source Cooperative Frontend & API](https://github.com/source-cooperative/source.coop) application. What's your reasoning for this?
4. Your comments in this PR suggest that the logs will not be written if the log level is set to `DEBUG` or `TRACE`, however you are calling the `log::debug` method which would indeed be logging when these conditions are true. Simple typo?

image

**Comment by kbgg** [comment]
Related: #5

File: src/apis/source.rs (hunk 1 above)
Comment (navdeepg2021) [reply src/apis/source.rs:473]: Yes It does output the Authorization header and that information is logged in cloudwatch logs. But as I said, this information is logged only if the log level is set to debug or trace.

**Comment by navdeepg2021** [comment]
> Hi Navdeep,
> 
> I have a couple of comments / concerns related to this PR:
> 
> 1. From a security standpoint, logging the entire request including Auth headers to the logs is not great.
> 2. Have you been able to reproduce this issue in your local environment first? That should be your first step in debugging this issue. This PR is labeled as a "fix" and I don't see how adding more logs fixes this issue.
> 3. Your debugging seems to be occurring in a different project than the errors are happening in. From the screenshot that Amit shared with me (attached below) the errors are happening in the Vercel which is the [Source Cooperative Frontend & API](https://github.com/source-cooperative/source.coop) application. What's your reasoning for this?
> 4. Your comments in this PR suggest that the logs will not be written if the log level is set to `DEBUG` or `TRACE`, however you are calling the `log::debug` method which would indeed be logging when these conditions are true. Simple typo?
> 
> image

I appreciate your help in resolving this issue. 

I wasn't able to reproduce this issue in my local and for some reason the access_key_id was being set to anonymous. I have seen lot of logs in vercel where the api-keys is called with anonymous access_key_id. I think when the calls are coming from UI, its coming as anonymous when the user is not logged in.

I have tried to reproduce this error in my local again and sometimes it does comes as empty as well resulting in the error that we saw in vercel.

The thinking behind adding the logs here was to find the call stack resulting in this error as since data proxy was found to be calling this method, we added extra logging here which along with the original caller would be a useful information to troubleshoot this further.

I saw that you have opened an issue about the use case but the original root cause of this error is still mystery. Probably there are calls being made to data proxy without valid credentials resulting in this error which can be filtered out well before they reach to the source API. What do you think about this?

**Comment by kbgg** [comment]
@navdeepg2021 I was able to reproduce this issue locally, have you tried setting `AWS_SECRET_ACCESS_KEY` to some value, like `foobar`? My inclination is that the AWS CLI will make anonymous requests when both `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are empty.

My assumption is that the root cause of this issue is someone making an S3 request to the data proxy with misconfigured credentials. This is not an unreasonable since this is a public API and this was (at least on Wednesday) an intermittent issue.

The data proxy also should be making use of a TTL cache in order to not hit the API so frequently, however there seems to be a case where when the cache key is set to an empty string (as would be the case when the access key id provided is empty) then there is always a cache miss. This could be fixed as well however adding a check for when the access key id is empty and bypassing the API call entirely would make this fix moot as well as reducing the total number of API calls made.

I'm pretty firmly against the debugging of this issue in prod. This fix would cause the `SOURCE_KEY` environment variable to be written to the CloudWatch logs if you are indeed attempting to debug this issue in prod. It is [extremely bad security practice to write sensitive information to logs](https://www.skyflow.com/post/how-to-keep-sensitive-data-out-of-your-logs-nine-best-practices).

Fixing the bug in #5 would resolve this issue.

cc @jedsundwall

**Pull request closed without merging**